### PR TITLE
Add progress tracking on Loops - 2/n

### LINKED ISSUE #6429: Formalize progress tracking inside of the trainer internals
## 🚀 Feature

We should better enforce progress tracking across these dimensions:
Stage: training, evaluation (validation and test), and prediction loops
Granularity: batches, steps, epochs
- batches vs steps: steps = optimizer steps (parameter updates) and applies to training loop only. this will differ from batches when gradient accumulation is used, for instance

### Motivation

- Provide consistency across trainer stages for users
- Better debugging
- Create mid-epoch resumable state for training

### Pitch

See the full example [here](https://gist.github.com/ananthsub/63c4c625af6a1b8cb45cfc8a5f52173d)

1. Create a dataclass that each Loop in Lightning is responsible for maintaining. All of this state is local to a particular trainer rank. These dataclasses live as attributes on the corresponding loop.
```
@dataclass
class LoopProgress:
    # this also serves as the index of the current epoch
    total_epochs_processed: int = 0
    # this monotonically increases and is summed across epochs
    total_batches_processed: int = 0
    # this resets at the end of the epoch back to 0
    batches_processed_this_epoch: int = 0
    
    # convenience utils for accessing state
    def bump_batch(self, increment: int = 1):
        self.total_batches_processed += increment
        self.batches_processed_this_epoch += increment
    
    def bump_epoch(self, increment: int = 1):
        self.total_epochs_processed += increment
    
    def reset_batch_in_epoch(self):
        self.batches_processed_this_epoch = 0
```

The train loop extends this to track optimizer steps
```
@dataclass
class TrainLoopProgress(LoopProgress):
    total_optimizer_steps_processed: int = 0
    optimizer_steps_processed_this_epoch: int = 0

    def bump_step(self, increment: int = 1):
        self.total_optimizer_steps_processed += increment
        self.optimizer_steps_processed_this_epoch += increment
    
    def reset_step_in_epoch(self):
        self.optimizer_steps_processed_this_epoch = 0
```

2. The `Trainer` maintains its own tracker that has references to the individual loop progress trackers
```
@dataclass
class Progress:
    train_progress: TrainLoopProgress
    val_progress: LoopProgress
    test_progress: LoopProgress
    predict_progress: LoopProgress
```

3. For convenience, we could offer synchronization utilities to sum the progress state across ranks to get totals
4. We can also offer convenience utilities to get the totals across different stages
5. We update the loops to populate this state, and make backwards compatible changes to reference this state
6. We save the progress state into checkpoints as part of the trainer state
7. We handle loading this state when resuming from checkpoints

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR build on top of https://github.com/PyTorchLightning/pytorch-lightning/pull/8334. Merge in order.

This PR adds progress tracking to Fit / Evaluation / Predict loops and ensures the correct progress state.

Fixes #6429

### Does your PR introduce any breaking changes? If yes, please list them.

No

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)
- [x] Did you list all the breaking changes introduced by this pull request?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/core/optimizer.py [hunk 1]
```diff
@@ -207,6 +207,7 @@ def closure_dis():
             profiler_name = f"optimizer_step_and_closure_{self._optimizer_idx}"
 
         self.__optimizer_step(*args, closure=closure, profiler_name=profiler_name, **kwargs)
+        self._trainer.fit_loop.epoch_loop.batch_loop.optim_progress.optimizer.step.increment_processed()
```
Comment (justusschock) [inline pytorch_lightning/core/optimizer.py:210]: This assumes a rigid loop structure. Is there maybe a way to do this more dynamically? Ideally the trainer (and optimizer, lightningmodule etc.) only know about the entry point to the top-most loop and everything else is handled by the loops.

thoughts on how to handle this @awaelchli @ananthsub @carmocca  ?

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 2]
```diff
@@ -421,13 +441,17 @@ def _optimizer_step(
             using_lbfgs=is_lbfgs,
         )
```
Comment (justusschock) [inline pytorch_lightning/loops/batch/training_batch_loop.py:443]: Could this be the place where we increment the progress? This would be the same, right? Or wouldn't this work for manual optimization?

File: pytorch_lightning/trainer/connectors/optimizer_connector.py [hunk 3]
```diff
@@ -48,6 +49,8 @@ def update_learning_rates(
         if opt_indices is None:
             opt_indices = []
 
+        progress: OptimizationProgress = self.trainer.fit_loop.epoch_loop.batch_loop.optim_progress
```
Comment (justusschock) [inline pytorch_lightning/trainer/connectors/optimizer_connector.py:52]: same as above. Is there a way to not hardcode how many nesting levels we actually have? Custom loops could either combine two loops into one or have more levels which would make this invalid.

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 4]
```diff
@@ -50,29 +50,17 @@ def __init__(self) -> None:
         self.running_loss: TensorRunningAccum = TensorRunningAccum(window_length=20)
         self.batch_idx: int = 0
         self.split_idx: Optional[int] = None
-        self.progress = BatchProgress()
+        self.progress = Progress()
```
Comment (justusschock) [inline pytorch_lightning/loops/batch/training_batch_loop.py:53]: @tchaton why did you change the type to the super type here?

File: pytorch_lightning/core/optimizer.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/core/optimizer.py:210]: Moved to batch_loop.

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/loops/batch/training_batch_loop.py:443]: Yes, moved it there, and yes it won't work for manual optimization.

Comment (tchaton) [reply pytorch_lightning/loops/batch/training_batch_loop.py:443]: We will handle manual optimization in another PR.

File: pytorch_lightning/trainer/connectors/optimizer_connector.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/optimizer_connector.py:52]: For now, we could do `self.trainer.fit_loop.epoch_loop.batch_loop.increment_scheduler_ready()`

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/loops/batch/training_batch_loop.py:53]: Reverted back to `BatchProgress `

File: pytorch_lightning/loops/base.py [hunk 5]
```diff
@@ -56,8 +56,8 @@ def trainer(self) -> Optional['pl.Trainer']:
 
     @trainer.setter
     def trainer(self, trainer: 'pl.Trainer'):
-        """Connect the Trainer to this loop and all children."""
-        if not isinstance(trainer, pl.Trainer) and trainer is not None:
+        """Connect the Trainer to itself and all its children loops"""
```
Comment (carmocca) [inline pytorch_lightning/loops/base.py:59]: "Connect the Trainer to itself" doesn't sound right

```suggestion
        """Connects this loop's trainer and it's children"""
```

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 6]
```diff
@@ -52,27 +52,15 @@ def __init__(self) -> None:
         self.split_idx: Optional[int] = None
         self.progress = BatchProgress()
         self.optim_progress = OptimizationProgress()
-
```
Comment (carmocca) [inline pytorch_lightning/loops/batch/training_batch_loop.py:55]: Keep this whitespace to split the public from the protected properties?

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 7]
```diff
@@ -33,10 +33,8 @@ class EvaluationLoop(DataLoaderLoop):
     def __init__(self):
         super().__init__()
         self.outputs = []
-        self.progress = EpochLoopProgress()
-
+        self.progress = DataLoaderProgress()
         self.epoch_loop = EvaluationEpochLoop()
-
```
Comment (carmocca) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:39]: Same whitespace comment

File: pytorch_lightning/loops/dataloader/prediction_loop.py [hunk 8]
```diff
@@ -20,9 +20,7 @@ def __init__(self):
         self.predictions: Optional[List[List[Any]]] = None
         self.epoch_batch_indices: Optional[List[List[int]]] = None
         self.progress = EpochLoopProgress()
-
         self.epoch_loop = PredictionEpochLoop()
-
```
Comment (carmocca) [inline pytorch_lightning/loops/dataloader/prediction_loop.py:25]: Same whitespace comment

File: tests/loops/test_loop_progress_integration.py [hunk 9]
```diff
@@ -1,22 +1,19 @@
...
     # check the validation progresses are not shared
     assert trainer.fit_loop.epoch_loop.val_loop.progress is not trainer.validate_loop.progress
+    expected = _collect_loop_progress(trainer.fit_loop)["epoch_loop"]["progress"]
+    assert expected == trainer.fit_loop.epoch_loop.progress
+    expected = _collect_loop_progress(trainer.fit_loop)["epoch_loop"]["batch_loop"]["progress"]
+    assert expected == trainer.fit_loop.epoch_loop.batch_loop.progress
+    expected = _collect_loop_progress(trainer.fit_loop)["epoch_loop"]["val_loop"]["progress"]
+    assert expected == trainer.fit_loop.epoch_loop.val_loop.progress
+    expected = _collect_loop_progress(trainer.fit_loop)["epoch_loop"]["val_loop"]["epoch_loop"]["progress"]
+    assert expected == trainer.fit_loop.epoch_loop.val_loop.epoch_loop.progress
```
Comment (carmocca) [inline tests/loops/test_loop_progress_integration.py:19]: This should avoid calling `_collect_loop_progress(trainer.fit_loop)` every time.

Also, the purpose of this test was to check that the references are correct, which are not there after your changes, so the whole test could be removed.

File: pytorch_lightning/trainer/trainer.py [hunk 10]
```diff
@@ -1010,6 +1010,7 @@ def _run_train(self) -> None:
                 self.training_type_plugin.reconciliate_processes(traceback.format_exc())
             # give accelerators a chance to finish
             self.accelerator.on_train_end()
+            self.on_expection()
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:1013]: As discussed somewhere else, this should not call this and instead the `on_train_end` hook should be called

File: pytorch_lightning/utilities/imports.py [hunk 11]
```diff
@@ -101,3 +102,7 @@ def _compare_version(package: str, op, version) -> bool:
     _IPU_AVAILABLE = poptorch.ipuHardwareIsAvailable()
 else:
     _IPU_AVAILABLE = False
+
+
+def fault_tolerant_enabled():
```
Comment (carmocca) [inline pytorch_lightning/utilities/imports.py:107]:
```suggestion
def _fault_tolerant_enabled():
```

File: pytorch_lightning/trainer/progress.py [hunk 12]
```diff
@@ -213,18 +234,17 @@ def reset_on_epoch(self) -> None:
...
         self.scheduler.load_state_dict(state_dict["scheduler"])
+        self.optimizer_idx = state_dict["optimizer_idx"]
 
 
 @dataclass
 class EpochLoopProgress(BaseProgress):
     """
     Tracks epoch loop progress.
     These counters are local to a trainer rank. By default, they are not globally synced across all ranks.
-
```
Comment (carmocca) [inline pytorch_lightning/trainer/progress.py:223]: Can you configure your linter to not remove these newlines? The `Args:` section should be separated

File: pytorch_lightning/trainer/progress.py [hunk 13]
```diff
@@ -237,42 +257,3 @@ def reset_on_epoch(self) -> None:
...
-    Args:
-        epoch: Tracks epochs progress.
-    """
-
-    epoch: TrainingEpochProgress = field(default_factory=TrainingEpochProgress)
-
-    def reset_on_epoch(self) -> None:
-        # do not reset `epoch.current` as it should track the number of epochs this `fit` call
-        self.epoch.reset_on_epoch()
-        self.epoch.optim.reset_on_epoch()
```
Comment (carmocca) [inline pytorch_lightning/trainer/progress.py:278]: So why are these removed?

**Review by carmocca — CHANGES_REQUESTED** [review]

File: pytorch_lightning/trainer/progress.py (hunk 13 above)
Comment (tchaton) [reply pytorch_lightning/trainer/progress.py:278]: The have not used anymore. They can be removed.

File: tests/loops/test_loop_progress_integration.py (hunk 9 above)
Comment (tchaton) [reply tests/loops/test_loop_progress_integration.py:19]: This is still testing the references are correct.

Comment (tchaton) [reply tests/loops/test_loop_progress_integration.py:19]: And yes, let's call _collect_loop_progress only once.

Comment (carmocca) [reply tests/loops/test_loop_progress_integration.py:19]: > This is still testing the references are correct.

No because before was using `is` but now it's using `==`

**Comment by awaelchli** [comment]
@carmocca I addressed some of your comments

File: pytorch_lightning/trainer/progress.py (hunk 13 above)
Comment (tchaton) [reply pytorch_lightning/trainer/progress.py:278]: The progress is owned by each loop and composition is done automatically by the base loop.

Comment (tchaton) [reply pytorch_lightning/trainer/progress.py:278]: There shouldn't be any form of assumption of the progress structure as it would be a blocker for Loop Customization.

File: pytorch_lightning/trainer/trainer.py (hunk 10 above)
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:1013]: I don't believe there is a `on_train_end` outside of the hook on the Trainer now. Mind sharing more thoughts there ?

File: tests/loops/test_loop_progress_integration.py (hunk 9 above)
Comment (tchaton) [reply tests/loops/test_loop_progress_integration.py:19]: Oh, we can replace by is. They should be same reference.

Comment (tchaton) [reply tests/loops/test_loop_progress_integration.py]: @carmocca Done !

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 14]
```diff
@@ -158,7 +163,10 @@ def on_advance_end(self):
         # -----------------------------------------
         # VALIDATE IF NEEDED + CHECKPOINT CALLBACK
         # -----------------------------------------
-        should_check_val = self._should_check_val_fx(self.iteration_count, self.is_last_batch)
+        self.progress.should_check_val = should_check_val = self._should_check_val_fx(
+            self.iteration_count, self.is_last_batch
+        )
```
Comment (awaelchli) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:168]: do we save the `should_check_val` to the state dict in case we resume a step in which we want to check val?

if so, doesn't the local variable `should_check_val` here get derived from the restored progress state anyway?
not sure I understand why we need to save it

File: pytorch_lightning/trainer/progress.py [hunk 15]
```diff
@@ -140,6 +140,32 @@ class BatchProgress(Progress):
     """
 
 
+@dataclass
+class TrainingEpochProgress(Progress):
+    """
+    Tracks the batch progress
```
Comment (awaelchli) [inline pytorch_lightning/trainer/progress.py:146]:
```suggestion
    Tracks the epoch progress
```

File: pytorch_lightning/trainer/progress.py [hunk 16]
```diff
@@ -151,7 +177,7 @@ class EpochProgress(Progress):
         current: Tracks the current epoch progress
         batch: Tracks batch progress.
     """
-
+    dataloader_idx: int = 0
```
Comment (awaelchli) [inline pytorch_lightning/trainer/progress.py:180]: we have a dataloader_idx here but also in the special DataLoaderProgress?
Is this intentional?

File: pytorch_lightning/trainer/trainer.py (hunk 10 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:1013]: @tchaton I think fault tolerant should save a special checkpoint, not through the ModelCheckpoint callback (which is anyway not guaranteed to be present) because it would have side effects within that callback state. 

I think it could have a special name so we know this one should be used to resume a faulty experiment.

For these reasons I strongly think that we should just call `trainer.save_checkpoint()` here. 

What are the arguments against?

File: pytorch_lightning/trainer/progress.py (hunk 16 above)
Comment (tchaton) [reply pytorch_lightning/trainer/progress.py:180]: Yes, some dataclasses there are legacy and need to be dropped.

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 4 above)
Comment (carmocca) [reply pytorch_lightning/loops/batch/training_batch_loop.py:53]: BatchProgress does not do anything. Can be added back if necessary later

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 17]
```diff
@@ -165,6 +166,8 @@ def restore_training_state(self) -> None:
 
         self.restore_optimizers_and_schedulers()
 
+        self.restore_loops()
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:169]: @awaelchli You defined just above the `restore_progress` function which includes "loops" in the comment. Should we use that one?

Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py]: I merged the two functions just now, have a look

File: pytorch_lightning/trainer/progress.py [hunk 18]
```diff
@@ -213,66 +217,4 @@ def reset_on_epoch(self) -> None:
...
-
-    epoch: EpochProgress = field(default_factory=EpochProgress)
-
-    def increment_epoch_completed(self) -> None:
-        self.epoch.increment_completed()
-        self.reset_on_epoch()
-
-    def reset_on_epoch(self) -> None:
-        self.epoch.reset_on_epoch()
-        self.epoch.current.reset()
```
Comment (carmocca) [inline pytorch_lightning/trainer/progress.py:236]: How do you plan to encode this logic now that these do not exist?

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 19]
```diff
@@ -131,6 +124,10 @@ def on_run_start(self, batch: Any, batch_idx: int, dataloader_idx: int):
         void(batch_idx, dataloader_idx)
         self._remaining_splits = list(enumerate(self._tbptt_split_batch(batch)))
 
+    def on_advance_start(self, *args: Any, **kwargs: Any) -> None:
+        super().on_advance_start(*args, **kwargs)
+        self.progress.increment_started()
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/loops/batch/training_batch_loop.py:129]: I think this is different here. 
`self.progress` is for the batch loop which iterates over batch splits. 

the one that you mean is in `training_epoch_loop` which iterates over batches in an epoch

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 20]
```diff
@@ -98,6 +87,8 @@ def run(self, batch: Any, batch_idx: int, dataloader_idx: int) -> AttributeDict:
             self._warning_cache.warn("train_dataloader yielded None. If this was on purpose, ignore this warning...")
             return AttributeDict(signal=0, training_step_output=[[]])
 
+        self.progress.increment_ready()
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/loops/batch/training_batch_loop.py:90]: see comment below

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 21]
```diff
@@ -158,6 +166,12 @@ def advance(self, batch, batch_idx, dataloader_idx):
             if result:
                 self.batch_outputs[0].append(result.training_step_output)
 
+        self.progress.increment_processed()
+
+    def on_advance_end(self) -> None:
+        super().on_advance_end()
+        self.progress.increment_completed()
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/loops/batch/training_batch_loop.py:173]: it looks like we don't have any progress tracking through advance() in the training_epoch loop. Something is off here. 

The training epoch loop should track the count of the individual batches

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/loops/batch/training_batch_loop.py:129]: yep. 
the progress of a loop should be defined by how many advances we have made. otherwise what are we even tracking? how many calls to run() were made? that wouldn't be useful at all, and counter intuitive.

**Comment by awaelchli** [comment]
It looks like the FitLoop has no progress tracking, but we should add it there too to track the current epoch.

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 14 above)
Comment (tchaton) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:168]: The logic to restart from it has not been added yet. We are just saving it for later PR.

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/batch/training_batch_loop.py:90]: I think the batch is ready when provided to the batch loop as it tracks the batch logic.

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 22]
```diff
@@ -407,13 +428,18 @@ def _optimizer_step(
             using_lbfgs=is_lbfgs,
         )
 
+        self.optim_progress.optimizer.step.increment_processed()
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/batch/training_batch_loop.py:431]: This should have break then. Let's raise an exception if None.

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 23]
```diff
@@ -240,6 +254,11 @@ def _training_step_and_backward_closure(
         result = self.training_step_and_backward(split_batch, batch_idx, opt_idx, optimizer, hiddens)
         if result is not None:
             return_result.update(result)
+
+            # this should be done only if result.loss exists and ``optimizer step`` is being run
+            if not self.should_accumulate():
+                self.optim_progress.optimizer.step.increment_started()
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/batch/training_batch_loop.py:260]: Yes. Sounds good to me.

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 24]
```diff
@@ -682,3 +710,9 @@ def _truncated_bptt_steps(self) -> int:
         if lightning_module.truncated_bptt_steps > 0:
             return lightning_module.truncated_bptt_steps
         return self.trainer.truncated_bptt_steps or 0
+
+    def increment_scheduler_ready(self):
+        self.optim_progress.scheduler.increment_ready()
+
+    def increment_scheduler_completed(self):
+        self.optim_progress.scheduler.increment_completed()
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/batch/training_batch_loop.py:718]: We could remove them if scheduler tracking was moved to epoch loop. 
This is needed to defined the BatchLoop Interface.

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 17 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py]: `progress` is legacy one for now. We should refactor the restoration when we get rid or not of global_step.

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 25]
```diff
@@ -96,25 +91,40 @@ def reset(self) -> None:
         if isinstance(self._max_batches, int):
             self._max_batches = [self._max_batches] * len(self.dataloaders)
 
+        if self.restarting:
+            self.iteration_count = self.progress.dataloader_idx
+            self.restarting = False
+        else:
+            self.iteration_count = 0
+            # reset batch / epoch progress tracking
+            self.progress.current.reset()
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:100]: Yes, we agree there.

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 26]
```diff
@@ -33,8 +33,7 @@ class EvaluationLoop(DataLoaderLoop):
     def __init__(self):
         super().__init__()
         self.outputs = []
-        self.progress = EpochLoopProgress()
-
+        self.progress = DataLoaderProgress()
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:36]: IMO, The users shouldn't go through the attributes path. You should just do `self.trainer.fit_loop.total_epoch_completed` .

File: pytorch_lightning/trainer/progress.py (hunk 18 above)
Comment (tchaton) [reply pytorch_lightning/trainer/progress.py:236]: It would be explicitly in the code or on the other dataclasses.

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 14 above)
Comment (carmocca) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:168]: I agree with @awaelchli that ideally we don't save it but know if the batch should check val afterwards by looking at the restored state. It's maybe worth waiting to add it if not used in this PR

File: pytorch_lightning/loops/dataloader/evaluation_loop.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:36]: Depends, if you are talking about naive PL users, just `trainer.total_epochs_completed`

But for power PL users (who subclass or create their custom loops), they should be able to go through these paths

File: pytorch_lightning/trainer/trainer.py (hunk 10 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py]: I added an explicit `trainer.save_checkpoint()` call there. 
Have a look and if it's ok I can add a small test.

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/loops/batch/training_batch_loop.py:90]: resolving this, as you said you both have discussed this already

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 27]
```diff
@@ -148,7 +145,18 @@ def advance(self, batch, batch_idx, dataloader_idx):
         self.trainer.logger_connector.on_train_split_start(batch_idx, split_idx, split_batch)
 
         if self.trainer.lightning_module.automatic_optimization:
-            for opt_idx, optimizer in self.get_active_optimizers(batch_idx):
+            active_optimizers = self.get_active_optimizers(batch_idx)
+            for opt_idx, optimizer in active_optimizers:
+
+                # handle optimization restart
+                if self.restarting:
+                    if len(active_optimizers) > 1 and opt_idx < self.progress.current.completed:
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/loops/batch/training_batch_loop.py:153]: I think the progress is also not correct right?
Shouldn't it be 
`self.optim_progress.optimizer_idx`?

File: pytorch_lightning/trainer/trainer.py [hunk 28]
```diff
@@ -1249,3 +1251,11 @@ def _log_device_info(self) -> None:
                 "IPU available but not used. Set the `ipus` flag in your trainer"
                 " `Trainer(ipus=8)` or script `--ipus=8`."
             )
+
+    def _on_expection(self):
+        if not self.is_global_zero or not _fault_tolerant_enabled():
+            return
+
+        # save a checkpoint for fault tolerant training
+        file_path = os.path.join(self.default_root_dir, ".pl_auto_save.ckpt")
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:1259]: Why the period before the name? to hide it from the users?

File: pytorch_lightning/trainer/trainer.py [hunk 29]
```diff
@@ -1010,6 +1011,7 @@ def _run_train(self) -> None:
                 self.training_type_plugin.reconciliate_processes(traceback.format_exc())
             # give accelerators a chance to finish
             self.accelerator.on_train_end()
+            self._on_expection()
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:1014]: Should we also do this for `KeyboardInterrupt`?

Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:1024]: Great question, I would prefer not until a user requests it.

File: pytorch_lightning/trainer/trainer.py (hunk 28 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:1269]: open to suggestion for naming. 
it depends how Lightning will use this checkpoint. Will it be fully automated resuming? Or does the user interact with this file to manually restart?

we may also need to provide a rotating filename

File: tests/loops/test_loops.py [hunk 30]
```diff
@@ -123,99 +111,377 @@ def advance(self, *args: Any, **kwargs: Any) -> None:
...
+    has_opt_stepped_in_be = accumulate_grad_batches == 1 or n_batches % accumulate_grad_batches != 0
+
+    nbe_total_zero_grad = (nbe_stepping_batches + has_leftover_accumulation_batches) * n_optimizers
+    if accumulate_grad_batches > 1:
+        # FIXME: ready or completed? 0 or stop_optimizer?
+        be_total_zero_grad = (
+            n_optimizers + (be_batches_ready // accumulate_grad_batches - (accumulate_grad_batches > 1)) *
+            (n_optimizers - 1) + 0
+        )
+        # be_total_zero_grad = be_epoch_batches_ready // accumulate_grad_batches * n_optimizers + 0
```
Comment (carmocca) [inline tests/loops/test_loops.py:374]: Resolving this piece of the test is the last missing part as far as I am aware

**Review by carmocca — APPROVED** [review]
I think everything is resolved now 🥵 

We can think about the fault-tolerant checkpoint design in a later PR.

**Review by awaelchli — APPROVED** [review]
big one!

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:

File: tests/loops/test_loops.py [hunk 31]
```diff
@@ -123,99 +111,375 @@ def advance(self, *args: Any, **kwargs: Any) -> None:
...
+    assert trainer.fit_loop.epoch_loop.val_loop.epoch_loop.batch_progress.state_dict() == expected
+
+
+@mock.patch.dict(os.environ, {"PL_FAULT_TOLERANT_TRAINING": "1"})
+@pytest.mark.parametrize("accumulate_grad_batches", (1, 2, 3))
+@pytest.mark.parametrize("n_optimizers", (1, 3, 5))
+@pytest.mark.parametrize("stop_epoch", (1, 2))
+@pytest.mark.parametrize("stop_batch", (1, 2))
+@pytest.mark.parametrize("stop_optimizer", (1, 2))
+def test_loop_state_on_exception(accumulate_grad_batches, stop_epoch, stop_batch, stop_optimizer, n_optimizers, tmpdir):
```
Comment (SeanNaren) [inline tests/loops/test_loops.py:301]: Just for anyone coming to this PR, @carmocca pointed me to these tests which gave me a lot of confidence that the behaviour is as expected. Very strong tests to ensure that the behaviour is correct!

**Review by SeanNaren — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py [hunk 32]
```diff
@@ -1259,3 +1261,10 @@ def _log_device_info(self) -> None:
                 "IPU available but not used. Set the `ipus` flag in your trainer"
                 " `Trainer(ipus=8)` or script `--ipus=8`."
             )
+
+    def _on_expection(self):
```
Comment (daniellepintz) [inline pytorch_lightning/trainer/trainer.py:1265]: I have a question about this function - is there a reason it is a function instead of a few lines in the exception handling? This function is only called once

This function seems a bit like a callback, but that's not what it is right? Still learning and understanding

Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:1265]: It was just meant as a stub for further improvements:

> We can think about the fault-tolerant checkpoint design in a later PR.

It's entirely fine to change it - although it wouldn't be a callback as it only needs to run on one hook: `on_exception`

Comment (daniellepintz) [reply pytorch_lightning/trainer/trainer.py:1265]: got it, thank you!!
